realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/MongoClientTest.kt

```diff
@@ -876,6 +885,386 @@ class MongoClientTest {
+                val filter2 = Document("_id", doc2Id.insertedId)
+
+                this.updateOne(filter2, doc2).get()
+                this.updateOne(filter1, doc1).get()
+            }
+        }
+    }
+
+    @Test
+    fun watchStreamAsynchronous() {
+        looperThread.runBlocking {
+            with(getCollectionInternal()) {
+                val insertedDocument = Document("watch", "1")
+                        .apply {
+                            this["num"] = 1
+                        }
+
+                val updatedDocument = Document("watch", "1")
+                        .apply {
+                            this["num"] = 2
+                        }
+
+
+                val watcher = this.watch()
+
+                var eventCount = 0
+                watcher.getAsync { it ->
+                    if (it.isSuccess) {
+                        it.get().let { changeEvent ->
+                            when (eventCount) {
+                                0 -> {
+                                    assertEquals(OperationType.INSERT, changeEvent.operationType)
+                                    assertDocumentEquals(insertedDocument, changeEvent.fullDocument!!)
+                                }
+                                1 -> {
+                                    assertEquals(OperationType.REPLACE, changeEvent.operationType)
+                                    assertDocumentEquals(updatedDocument, changeEvent.fullDocument!!)
+                                }
+                                2 -> {
+                                    assertEquals(OperationType.DELETE, changeEvent.operationType)
+                                    assertNull(changeEvent.fullDocument)
+
+                                    watcher.cancel()
+                                }
+                            }
+                        }
+
+                        eventCount++
+                    } else {
+                        when (it.error.errorCode) {
+                            ErrorCode.NETWORK_IO_EXCEPTION -> looperThread.testComplete()
+                            else -> fail()
+                        }
+                        looperThread.testComplete()
+                    }
+                }
+
+                // Busy wait till watcher is ready to receive updates.
+                // It syncs the event producer thread (current thread) with
+                // the event consumer thread.
+                while (!watcher.isOpen) {
+                }
+
+                this.insertOne(insertedDocument).get()
+
+                val filter = Document("watch", "1")
+                this.updateOne(filter, updatedDocument).get()
+                this.deleteOne(filter).get()
+            }
+        }
+    }
+
+    @Test
+    fun watchStreamCancelSynchronous() {
+        looperThread.runBlocking {
+            with(getCollectionInternal()) {
+                val watcher = this.watch()
+
+                thread {
```

Comment:
I don't see it 😯! We are testing what happens when we cancel a watcher once it has been started:

1. Create watcher
2. On a new thread: Get next event
3. Wait for the watcher to initialize
4. Cancel the watcher (this will force the `getNextEvent` function to throw an exception)